error when chaining a function returning a record type within a rule guard

Chaining a function returning a record type within a rule guard causes the following error:

```
checker.c: In function ‘guard0’:
checker.c:4177:127: error: ‘ret30’ undeclared (first use in this function)
 4177 |   return ({ ru_consume(rule_name, state_drop_const(s), ({ ru_produce(rule_name, state_drop_const(s), (struct handle){ .base = ret30, .offset = 0ul, .width = 2ull }); })); }) ? 1 : 0;
      |                                                                                                                               ^~~~~
checker.c:4177:127: note: each undeclared identifier is reported only once for each function it appears in
```

Here is the model which results in the above error:

<details>
<summary>model for reproduction</summary>
<br>

```
type
  x_t: record
    y : boolean;
  end;

var
  z: boolean;

function produce(): x_t; 
var x: x_t;
begin
  x.y := true;
  return x;
end;

function consume(x: x_t): boolean;
begin
  return x.y;
end;

startstate begin
  z := true;
end;

rule consume(produce()) ==> begin
  z := !z;
end;
```
</details>

I used rumur version v2025.01.05 & gcc 11.4.0

A workaround, which works for my use case:
Instead of passing the record as an argument, I am calling the producer function from inside the consumer function:

<details>
<summary>workaround model</summary>
<br>

```
type
  x_t: record
    y : boolean;
  end;

var
  z: boolean;

function produce(): x_t; 
var x: x_t;
begin
  x.y := true;
  return x;
end;

function consume(): boolean;
var x: x_t;
begin
  x := produce();
  return x.y;
end;

startstate begin
  z := true;
end;

rule consume() ==> begin
  z := !z;
end;
```
</details>

Thank you for maintaining this project; it has been very useful thus far.

--- a/rumur/src/generate-allocations.cc
+++ b/rumur/src/generate-allocations.cc
@@ -39,6 +39,11 @@ private:
 
 } // namespace
 
+void generate_allocations(std::ostream &out, const Expr &expr) {
+  Generator g(out);
+  g.dispatch(expr);
+}
+
 void generate_allocations(std::ostream &out, const Stmt &stmt) {
   Generator g(out);
   g.dispatch(stmt);

--- a/rumur/src/generate-model.cc
+++ b/rumur/src/generate-model.cc
@@ -212,6 +212,10 @@ void generate_model(std::ostream &out, const Model &m) {
             out << ";\n";
           }
 
+          // allocate memory for any complex-returning functions we call
+          if (s->guard != nullptr)
+            generate_allocations(out, *s->guard);
+
           out << "  return ";
           if (s->guard == nullptr) {
             out << "true";

--- a/rumur/src/generate.h
+++ b/rumur/src/generate.h
@@ -13,6 +13,7 @@ int output_checker(const std::string &path, const rumur::Model &model,
                    const std::pair<ValueType, ValueType> &value_types);
 
 // Generate prelude definitions to allocate memory for function returns
+void generate_allocations(std::ostream &out, const rumur::Expr &expr);
 void generate_allocations(std::ostream &out, const rumur::Stmt &stmt);
 
 // Helper for calling the above on a body of functions

--- a/tests/return-record-in-guard.m
+++ b/tests/return-record-in-guard.m
@@ -1,0 +1,31 @@
+-- A function returning a record that is then used within a guard. This
+-- previously resulted in malformed code being emitted.
+-- https://github.com/Smattr/rumur/issues/290
+
+type
+  x_t: record
+    y : boolean;
+  end;
+
+var
+  z: boolean;
+
+function produce(): x_t; 
+var x: x_t;
+begin
+  x.y := true;
+  return x;
+end;
+
+function consume(x: x_t): boolean;
+begin
+  return x.y;
+end;
+
+startstate begin
+  z := true;
+end;
+
+rule consume(produce()) ==> begin
+  z := !z;
+end;

